### ISSUE TITLE
Behavior with giant source files

### ISSUE THREAD

**Issue by marijnh** [tree-sitter/tree-sitter#222]
In an editor context, if the user decides to open a ridiculously huge file (say, hundreds of megabytes of generated code or JSON data), I can imagine that even though the initial parse is acceptably fast, and subsequent updates cheap, the memory pressure from the giant syntax tree would be a problem. Is there a mechanism that could help mitigate this (like only keeping part of the syntax tree around), or is the recommended approach in such a situation to just not run the parser?

**Comment by maxbrunsfeld** [comment]
Yeah, this is a good question; I haven't really tackled this problem yet. The biggest files that I usually test with are 8-10MB generated C files.

I think you're right that the memory usage is fundamentally the limiting factor. I did [some work](https://github.com/tree-sitter/tree-sitter/pull/202) on memory usage recently, and after those optimizations, syntax trees still use over 10x as much memory as the size of the source file.

> Is there a mechanism that could help mitigate this (like only keeping part of the syntax tree around)

I think this would be a great addition. I can imagine several related options:

* Add an API for specifying a *region of interest* in the syntax tree, and discard parts of the syntax tree that are outside the region of interest. When updating the region of interest, you'd need to provide the source code again, in case some code needs to be re-parsed.
* Add a *large file mode*, in which we don't actually parse the entire source file. Instead, we could parse only a few hundred lines at a time, as needed, and rely on error recovery to obtain reasonable results despite missing context.
* Add an API that allows the syntax tree to read and write from disk, so that it can automatically persist inactive parts of the tree to disk and retrieve them on demand.

It'll take a bit of design work to decide on a path forward, so for the time being, the recommendation is to disable parsing altogether for files above some size threshold.

**Comment by marijnh** [comment]
Thanks for giving some background! I'd think for many kinds of functionality (such as go-to-definition or autocompletion) having half a syntax tree isn't going to help much (the feature will be unreliable, which can be worse than having it entirely disabled), but for syntax highlighting it'd suffice. Does the interface exposed to tools provide a way to tell them there's no syntax tree or only a partial syntax tree?

(Is asking questions on the bug tracker here considered annoying? I have a bunch more, but if there's a more suitable channel for that I'd be happy to use that.)

**Comment by maxbrunsfeld** [comment]
> Thanks for giving some background! I'd think for many kinds of functionality (such as go-to-definition or autocompletion) having half a syntax tree isn't going to help much (the feature will be unreliable, which can be worse than having it entirely disabled), but for syntax highlighting it'd suffice.

Yeah, I agree. I think that ideally, we'd display something in the UI that indicates that the editor's functionality is reduced due to the file size.

Does the interface exposed to Atom tools provide a way to tell them there's no syntax tree or only a partial syntax tree?

Right now, the APIs are still evolving. Because we're in the process of transitioning from TextMate parsers, some editors have a syntax tree API available, and some don't. In Atom, you can do this feature detection like this:

```js
const buffer = atom.workspace.getActiveTextEditor().getBuffer();
const languageMode = buffer.getLanguageMode();
if (languageMode.tree) {
  // This is a TreeSitterLanguageMode, so it has methods like `.getSyntaxNodeAtPosition()`.
} else {
  // This is a TextMateLanguageMode, so it does not have these APIs.
}
```

I'd like to take the time to clarify the APIs and document them better, but haven't done it yet. If we were to add some kind of degraded, large file mode, the feature detection APIs might become even more complicated 😬 .

**Comment by monnier** [comment]
>     * Add an API for specifying a _region of interest_ in the syntax tree, and discard parts of the syntax tree that are outside the region of interest. When updating the region of interest, you'd need to provide the source code again, in case some code needs to be re-parsed.

Indeed, I think a good approach is for tree-sitter to "fold" subtrees, i.e. throw away parts of the parse-tree and replace them with just enough info to reproduce them on demand (e.g. their *start* and *end* position in the source code, along with something like the corresponding production rule or the state of the engine at *start*).  In order to know what to fold, it could keep track of activity, or the application could inform it which region(s) are currently of interest (e.g. currently displayed).
